src/Xamarin.Android.Build.Tasks/Utilities/ManifestDocument.cs

```diff
@@ -839,11 +854,40 @@ public void SetAbi (string abi)
 			int code = 1;
 			if (!string.IsNullOrEmpty (VersionCode)) {
 				code = Convert.ToInt32 (VersionCode);
-				if (code > 0xffff || code < 0)
-					throw new ArgumentOutOfRangeException ("VersionCode", "VersionCode is outside 0, 65535 interval");
+				if (code > maxVersionCode || code < 0)
+					throw new ArgumentOutOfRangeException ("VersionCode", $"VersionCode is outside 0, {maxVersionCode} interval");
 			}
 			code |= GetAbiCode (abi) << 16;
 			VersionCode = code.ToString ();
 		}
+
+		public void CalculateVersionCode (string currentAbi, string versionCodePattern, string versionCodeProperties)
+		{
+			var regex = new Regex ("\\{(?<key>([A-Za-z]*)):?0*[\\}]");
```

Comment:
The problem is that there are multiple format strings which result in an integer, e.g. `g`, and I don't think that whitelisting is sane.

I'm also not sure that it matters, necessarily? If they specify e.g. `x4` and get a letter in the resulting versionCode, then `aapt`/etc. will complain later about the invalid version. So long as the resulting error message is sensible, it shouldn't matter *where* it errors, so long as it does.